Feature/datateam 255/improve docs and typing

nelsonwmoore opened the pull request:
## Standardize formatting, type hints, and docstrings for bento-meta
- finishes replacing util.cypher with minicypher
- add type hints throughout (WIP but almost entirely covered now)
- standardize docstrings to Google style
- formatting

nelsonwmoore committed:
style(*): formatting

nelsonwmoore committed:
fix(*): writeable mdb circular import

nelsonwmoore committed:
feat(minicypher): finish replacing util.cypher

Finish replacing util.cypher imports with minicypher throughout (util._engine.py, makeq.py, cypher tests and makeq tests)

nelsonwmoore committed:
style(tests): format tests

nelsonwmoore committed:
refactor(*): add type hints and docstrings (WIP)

nelsonwmoore committed:
refactor(*): address some more type errors

Add property setters to create data descriptors, enabling type checkers to recognize the getter's return type annotations.

nelsonwmoore committed:
refactor(loaders.py): add type hints

nelsonwmoore committed:
refactor(mdbs): add type hints

Decorators like `read_txn` are a little complicated to type hint. For now elected to hint the returns for the callers with what the query would return rather than what the decorator expects so that users have a better idea of what would be returned at runtime.

nelsonwmoore committed:
refactor(mdb_tools): type hints and doc strings

nelsonwmoore committed:
refactor(objects): add type hints

nelsonwmoore committed:
docs(*): update docstrings to google style

Update docstrings throughout bento-meta to consistently use Google style docstrings. Sphinx will still build docs automatically with no changes needed because we already use the 'napoleon' sphinx extension, but this style is a little more readable for human devs.

nelsonwmoore committed:
chore: import LiteralString from typing_extensions

To support Python 3.10

nelsonwmoore committed:
fix: remaining typing.LiteralString import

nelsonwmoore committed:
fix(*): issues identified by copilot

- entity.py: suggested fix didn't work because CollValue.__setitem__() expects an entity value. Instead implement `belongs` tracking for the Entity owner & mark as dirty, which achieves the desired effect of triggering __setitem__
- object_map.py: use elif rather than if when checking CollValue values type
- model.py: use del to remove item from prop owner's collection rather than setting value to None

nelsonwmoore commented on python/src/bento_meta/model.py:
```diff
@@ -111,239 +143,301 @@ def add_node(self, node=None):
             owner = prop.belongs[okey]
             (i, att, key) = okey
-            getattr(owner, att)[key] == None
+            getattr(owner, att)[key] = None
```
addressed in https://github.com/CBIIT/bento-meta/pull/72/commits/5d069082d58f0b844016f88fb83274bafa9c6cc1

nelsonwmoore commented on python/src/bento_meta/object_map.py:
```diff
@@ -434,205 +535,166 @@ def put_q(self, obj):
         if isinstance(values, CollValue):
             values = values.values()
-        elif isinstance(values, Entity):
+        if isinstance(values, Entity):
```
addressed in https://github.com/CBIIT/bento-meta/pull/72/commits/5d069082d58f0b844016f88fb83274bafa9c6cc1

nelsonwmoore commented on python/src/bento_meta/entity.py:
```diff
@@ -457,30 +522,33 @@ def owner_key(self) -> str:
 
     def __delitem__(self, name: str) -> None:
+        """Delete the value for the collection."""
         self[name] == None  # trigger __setitem__
```
addressed in https://github.com/CBIIT/bento-meta/pull/72/commits/5d069082d58f0b844016f88fb83274bafa9c6cc1

knockknockyoo commented on python/tests/test_013mdbtools.py:
```diff
@@ -186,7 +202,7 @@ def test_get_relationship_between_entities(self, tools_mdb) -> None:
                 R(Type=rel_type),
                 N(label=self.node_2.get_label(), props=self.node_2.get_attr_dict()),
-            )
+            ),
```
Can you double check unnecessary changes? I dont think we need a comma.

nelsonwmoore replied:
Sure, these were auto-added by the formatter but I agree they're not necessary in these cases where we only have one argument. I removed the instances you pointed out and some others I found.

knockknockyoo commented on python/tests/test_013mdbtools.py:
```diff
@@ -237,7 +253,7 @@ def test_term_validation_success():
             "origin_version": "test_version",
             "value": "test_value",
-        }
+        },
```
unnecessary changes

nelsonwmoore replied:
removed comma

knockknockyoo commented on python/tests/test_013mdbtools.py:
```diff
@@ -266,7 +282,7 @@ def test_validation_failure_edge_src_dst_attr_missing():
             "src": Node({"model": "test_model"}),
             "dst": Node({"handle": "destination"}),
-        }
+        },
```
please remove comma

nelsonwmoore replied:
removed comma

knockknockyoo commented on python/tests/test_013mdbtools.py:
```diff
@@ -46,15 +49,15 @@ class TestToolsMDB:
     pg_valset_2 = N(label=valset_2.get_label(), props=valset_2.get_attr_dict())
     term_1 = Term(
-        {"value": "term_1", "origin_name": "origin_name_1", "nanoid": "tnano1"}
+        {"value": "term_1", "origin_name": "origin_name_1", "nanoid": "tnano1"},
```
please remove comma

nelsonwmoore replied:
ok

nelsonwmoore committed:
fix(*): remove unnecessary trailing commas

remove unnecessary trailing from single-argument cases. keep them for multi-item collections that span multiple lines

nelsonwmoore commented:
Latest commits showing up on branch but PR not picking them up. closing & reopening to try to force a sync

knockknockyoo reviewed (approved):
Approved.
